File: matlab/src/lang/nodes/ClassSwitch.m
classdef ClassSwitch < Node
    % A node to change the class of visiting jobs
    %
    % Copyright (c) 2012-2026, Imperial College London
    % All rights reserved.

    properties (Hidden)
        autoAdded;
    end

    properties
        cap;
        schedPolicy;
        schedStrategy;
    end

    methods
        %Constructor
        function self = ClassSwitch(model, name, csMatrix)
            % SELF = CLASSSWITCH(MODEL, NAME, CSMATRIX)
            
            self@Node(name);
            if isa(model,'Network')
                % Handle the new delegation pattern
                if model.isMatlabNative()
                    classes = model.getClasses();
                    if nargin < 3
                        csMatrix = eye(length(classes));
                    end
                    
                    self.autoAdded = false;
                    self.input = Buffer(classes);
                    self.output = Dispatcher(classes);
                    self.cap = Inf;
                    self.schedPolicy = SchedStrategyType.NP;
                    self.schedStrategy = SchedStrategy.FCFS;
                    self.server = StatelessClassSwitcher(classes, csMatrix);
                    self.setModel(model);
                    model.addNode(self);
                else
                    % Java implementation through delegation
                    self.setModel(model);
                    if nargin < 3
                        self.obj = jline.lang.nodes.ClassSwitch(model.implementation.obj, name);
                    else
                        self.obj = jline.lang.nodes.ClassSwitch(model.implementation.obj, name, csMatrix);
                    end
                    self.index = model.implementation.obj.getNodeIndex(self.obj);
                end
            elseif isa(model,'MNetwork') || (isa(model,'Network') && model.isMatlabNative())
                classes = model.getClasses();
                if nargin < 3
                    csMatrix = eye(length(classes));
                end
                
                self.autoAdded = false;
                self.input = Buffer(classes);
                self.output = Dispatcher(classes);
                self.cap = Inf;
                self.schedPolicy = SchedStrategyType.NP;
                self.schedStrategy = SchedStrategy.FCFS;
                self.server = StatelessClassSwitcher(classes, csMatrix);
                self.setModel(model);
                self.model.addNode(self);
            elseif isa(model,'JNetwork') || (isa(model,'Network') && model.isJavaNative())
                self.setModel(model);
                if nargin < 3
                    self.obj = jline.lang.nodes.ClassSwitch(model.obj, name);
                else
                    self.obj = jline.lang.nodes.ClassSwitch(model.obj, name, csMatrix);
                end
                self.index = model.obj.getNodeIndex(self.obj);
            end
        end

        function C = initClassSwitchMatrix(self)
            % C = INITCLASSSWITCHMATRIX()

            K = self.model.getNumberOfClasses;
            C = zeros(K,K);
        end

        function setClassSwitchingMatrix(self, csMatrix)
            self.server.updateClasses(self.model.getClasses());
            self.server.updateClassSwitch(csMatrix);
        end

        function setProbRouting(self, class, destination, probability)
            % SETPROBROUTING(CLASS, DESTINATION, PROBABILITY)

            setRouting(self, class, RoutingStrategy.PROB, destination, probability);
        end

        function summary(self)
            % SUMMARY()

            line_printf('\nNode: <strong>%s</strong>',self.getName);
            for r=1:length(self.output.outputStrategy)
                %classes = self.model.getClasses();
            %line_printf('Routing %s: %s',classes{r}.name,self.output.outputStrategy{r}{2});
                for s=1:length(self.output.outputStrategy)
                    if self.server.csMatrix(r,s)>0
                        classes = self.model.getClasses();
            line_printf('Routing %s->%s: %g',classes{r}.name,classes{s}.name,self.server.csMatrix(r,s));
                    end
                end
            end
            %            self.input.summary;
            %            self.server.summary;
            %            self.output.summary;
        end
    end

end

File: jar/src/main/kotlin/jline/lang/nodes/ClassSwitch.java
/*
 * Copyright (c) 2012-2026, QORE Lab, Imperial College London
 * All rights reserved.
 */

package jline.lang.nodes;

import jline.lang.ClassSwitchMatrix;
import jline.lang.JobClass;
import jline.lang.Network;
import jline.lang.constant.RoutingStrategy;
import jline.lang.constant.SchedStrategy;
import jline.lang.constant.SchedStrategyType;
import jline.lang.sections.Buffer;
import jline.lang.sections.Dispatcher;
import jline.lang.sections.StatelessClassSwitcher;
import jline.util.matrix.Matrix;

import java.io.Serializable;
import java.util.List;

/**
 * A node that switches the class of an incoming job based on a probability table
 */
public class ClassSwitch extends Node implements Serializable {
    public boolean autoAdded;
    protected SchedStrategyType schedPolicy;
    protected SchedStrategy schedStrategy;

    public ClassSwitch(Network model, String name) {
        super(name);

        List<JobClass> jobClasses = model.getClasses();
        int m = jobClasses.size();

        this.input = new Buffer(jobClasses);
        this.output = new Dispatcher(jobClasses);

        Matrix csFun = new Matrix(m, m, m * m);

        this.setModel(model);
        this.model.addNode(this);

        this.schedPolicy = SchedStrategyType.NP;
        this.schedStrategy = SchedStrategy.FCFS;
        this.server = new StatelessClassSwitcher(jobClasses, csFun);
        this.autoAdded = false;
    }

    public ClassSwitch(Network model, String name, Matrix csFun) {
        super(name);

        List<JobClass> jobClasses = model.getClasses();
        this.input = new Buffer(jobClasses);
        this.output = new Dispatcher(jobClasses);


        this.setModel(model);
        this.model.addNode(this);

        this.schedPolicy = SchedStrategyType.NP;
        this.schedStrategy = SchedStrategy.FCFS;
        this.server = new StatelessClassSwitcher(jobClasses, csFun);
    }

    public ClassSwitchMatrix initClassSwitchMatrix() {
        int m = this.model.getNumberOfClasses();
        // Find the maximum class index to ensure matrix can accommodate 1-based indexing
        int maxIndex = 0;
        for (JobClass jc : this.model.getClasses()) {
            maxIndex = Math.max(maxIndex, jc.getIndex());
        }
        // Matrix size needs to be at least maxIndex+1 to accommodate 0-based access with 1-based indices
        int matrixSize = Math.max(m, maxIndex + 1);
        return new ClassSwitchMatrix(matrixSize, matrixSize, matrixSize * matrixSize);
    }

    public void setClassSwitchingMatrix(ClassSwitchMatrix csMatrix) {
        ((StatelessClassSwitcher) this.server).updateClasses(this.model.getClasses());
        ((StatelessClassSwitcher) this.server).updateClassSwitch(csMatrix.getMatrix());
    }

    public void setProbRouting(JobClass jobClass, Node destination, double probability) {
        this.setRouting(jobClass, RoutingStrategy.PROB, destination, probability);
    }

}
